src/Driver/ResumableJsUploadDriver.php

```diff
@@ -0,0 +1,193 @@
+     */
+    public function resume(Request $request, StorageConfig $config): Response
+    {
+        $this->validateChunkRequest($request);
+
+        try {
+            $range = new OneBasedRequestBodyRange(
+                $request->query,
+                $this->buildParameterName('chunk-number'),
+                $this->buildParameterName('total-chunks'),
+                $this->buildParameterName('chunk-size'),
+                $this->buildParameterName('total-size')
+            );
+        } catch (InvalidArgumentException $e) {
+            throw new BadRequestHttpException($e->getMessage(), $e);
+        }
+
+        $filename = $request->query($this->buildParameterName('identifier'));
+        $chunkname = $this->buildChunkname($range);
+
+        if (! $this->chunkExists($config, $filename, $chunkname)) {
+            throw new NotFoundHttpException();
+        }
+
+        return new JsonResponse(['OK']);
+    }
+
+    /**
+     * @param \Illuminate\Http\Request $request
+     * @param StorageConfig $config
+     * @param \Closure|null $fileUploaded
+     *
+     * @return \Symfony\Component\HttpFoundation\Response
+     */
+    public function save(Request $request, StorageConfig $config, Closure $fileUploaded = null): Response
+    {
+        $file = $request->file($this->fileParam);
+
+        $this->validateUploadedFile($file);
+
+        $this->validateChunkRequest($request);
+
+        return $this->saveChunk($file, $request, $config, $fileUploaded);
+    }
+
+    /**
+     * @param Request $request
+     */
+    private function validateChunkRequest(Request $request)
+    {
+        $validation = [];
+
+        foreach ($this->parameterNames as $key => $_) {
+            $validation[$this->buildParameterName($key)] = 'required';
+        }
+
+        $request->validate($validation);
+    }
+
+    /**
+     * @param UploadedFile $file
+     * @param Request $request
+     * @param StorageConfig $config
+     * @param \Closure|null $fileUploaded
+     *
+     * @return Response
+     */
+    private function saveChunk(UploadedFile $file, Request $request, StorageConfig $config, Closure $fileUploaded = null): Response
+    {
+        try {
+            $range = new OneBasedRequestBodyRange(
+                $request,
+                $this->buildParameterName('chunk-number'),
+                $this->buildParameterName('total-chunks'),
+                $this->buildParameterName('chunk-size'),
+                $this->buildParameterName('total-size')
+            );
+        } catch (InvalidArgumentException $e) {
+            throw new BadRequestHttpException($e->getMessage(), $e);
+        }
+
+        $filename = $request->post($this->buildParameterName('identifier'));
+
+        $chunks = $this->storeChunk($config, $range, $file, $filename);
+
+        if (!$range->isFinished($chunks)) {
+            return new PercentageJsonResponse($range->getPercentage($chunks));
+        }
+
+        $targetFilename = $file->hashName();
```

Comment:
At first this could seem as a "useless" variable because it is only used once. However, the variable name it self contains information about the value it has.